File: dp/leetcode_1137.java
package dp;

public class leetcode_1137 {

    static int[] tribonacciArr = new int[38];

    public static void main(String[] args) {
        leetcode_1137 lc = new leetcode_1137();
        System.out.println(lc.tribonacci(4));
    }

    public int tribonacci(int n) {
        if (n == 0) return 0;
        if (n <= 2) return 1;

        if (tribonacciArr[n] == 0) {
            tribonacciArr[n] = tribonacci(n - 1) + tribonacci(n - 2) + tribonacci(n - 3);
        }

        return tribonacciArr[n];
    }
}

File: dp/LeetCode_1137.kt
package dp

class LeetCode_1137 {

    fun tribonacci(n: Int): Int {
        if (n < 2) {
            return n
        }

        val numQueue = intArrayOf(0, 0, 1)

        var lastN = 1
        var pointer = 0

        repeat(n - 2) {
            val pointedValue = numQueue[pointer]
            numQueue[pointer] = lastN
            lastN = lastN * 2 - pointedValue
            pointer = if (pointer + 1 in numQueue.indices) pointer + 1 else 0
        }

        return lastN
    }
}

fun main() {
    // Input
    val n = 4
    // Output
    LeetCode_1137().tribonacci(n).let {
        println(it)
        require(it == 4)
    }
}
